### PR TITLE
Laithhanania user auth

### PR DESCRIPTION
**Pull request by LaithHanania**
Implemented authentication for the react application.

Implementation flow for first login:
1. User clicks login
2. Taken to Github login page
3. On sign in, user is routed back to the application with a code in the callback URL
4. code is taken from callback URL and sent to the backend
5. Code is exchanged for Github access token
6. Github access token is used to get user details from Github and the user in MongoDB is created / updated accordingly
7. JWT access token created based on username
8. Access token and username sent back to frontend, user is now authenticated
9. User sent to back to frontend and stored in local storage and recoil state

Implementation flow for subsequent reloads:
1. on app render the access token in the local storage is checked for validity, if invalid it is deleted and user is signed out. To sign back in the aforementioned flow is used
2. If access token is valid, then the associated user is retrieved and sent back to the frontend and user stored in recoil state. Subsequent requests that require authentication must use bearer authorization headers with the access token to be validated.

Also implemented login modal and logout on the frontend.

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -18,17 +23,11 @@
 from config import config
 
 bootstrap = Bootstrap()
-db = MongoEngine()
```
Comment (LaithHanania) [inline app/__init__.py:21]: relocated to db.py to avoid circular import

File: app/__init__.py [hunk 2]
```diff
@@ -18,17 +23,11 @@
...
 mail = Mail()
 github = GitHub()
-login_manager = LoginManager()
```
Comment (LaithHanania) [inline app/__init__.py:24]: relocated to loginmanager.py to avoid circular import

File: app/models.py [hunk 3]
```diff
@@ -85,30 +85,30 @@ class User(UserMixin, db.Document):
     permission = db.IntField()
     last_seen = db.DateTimeField()
     followed_projects = db.ListField(db.StringField())
-    followed_forks = db.ListField(db.StringField()) # not used
+    followed_forks = db.ListField(db.StringField())  # not used
```
Comment (LaithHanania) [inline app/models.py:88]: python Black formatter automatically made these stylistic changes

File: client/src/App.js [hunk 4]
```diff
@@ -21,24 +25,47 @@ const theme = createTheme({
...
-              element={
-                !isEmpty(currentUser) ? (
-                  <Navigate to="/followed"/>
-                ) : (
-                  <Navigate to="/login" />
-                )
-              } 
-            />
-            <Route path="/login" element={<div>please login </div>} />
+            <Route path="/" element={<Home />} />
```
Comment (LaithHanania) [inline client/src/App.js:67]: Switched base path to a home component as we will implement a welcome page in the future and a base path is required for authentication

File: client/src/Home.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
```
Comment (LaithHanania) [inline client/src/Home.js:1]: will be replaced with a welcome component in a coming ticket

**Review by johnsoncv — APPROVED** [review]
lgtm!

**Review by johnperry99 — APPROVED** [review]
LGTM

**Pull request merged**